analyzers/src/SonarAnalyzer.CSharp/Rules/UseConstantLoggingTemplate.cs

```diff
@@ -0,0 +1,125 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class UseConstantLoggingTemplate : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S2629";
+    private const string MessageFormat = "{0}";
+    private const string OnUsingStringInterpolation = "Don't use string interpolation in logging message templates.";
+    private const string OnUsingStringFormat = "Don't use String.Format in logging message templates.";
+    private const string OnUsingStringConcatenation = "Don't use string concatenation in logging message templates.";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    private static readonly ImmutableDictionary<SyntaxKind, string> Messages = new Dictionary<SyntaxKind, string>
+    {
+        {SyntaxKind.AddExpression, OnUsingStringConcatenation},
+        {SyntaxKind.InterpolatedStringExpression, OnUsingStringInterpolation},
+        {SyntaxKind.InvocationExpression, OnUsingStringFormat},
+    }.ToImmutableDictionary();
+
+    private static readonly ImmutableArray<KnownType> LoggerTypes = ImmutableArray.Create(
+        KnownType.Castle_Core_Logging_ILogger,
+        KnownType.log4net_ILog,
+        KnownType.log4net_Util_ILogExtensions,
+        KnownType.Microsoft_Extensions_Logging_LoggerExtensions,
+        KnownType.NLog_ILogger,
+        KnownType.NLog_ILoggerBase,
+        KnownType.NLog_ILoggerExtensions,
+        KnownType.Serilog_ILogger,
+        KnownType.Serilog_Log);
+
+    private static readonly ImmutableHashSet<string> LoggerMethodNames = ImmutableHashSet.Create(
+        "ConditionalDebug",
+        "ConditionalTrace",
+        "Debug",
+        "DebugFormat",
+        "Error",
+        "ErrorFormat",
+        "Fatal",
+        "FatalFormat",
+        "Info",
+        "InfoFormat",
+        "Information",
+        "Log",
+        "LogCritical",
+        "LogDebug",
+        "LogError",
+        "LogFormat",
+        "LogInformation",
+        "LogTrace",
+        "LogWarning",
+        "Trace",
+        "TraceFormat",
+        "Verbose",
+        "Warn",
+        "WarnFormat",
+        "Warning");
+
+    private static readonly ImmutableHashSet<string> LogMessageParameterNames = ImmutableHashSet.Create(
+        "format",
+        "message",
+        "messageTemplate");
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+        {
+            var invocation = (InvocationExpressionSyntax)c.Node;
+            if (LoggerMethodNames.Any(x => x == invocation.GetName())
```

Comment:
```suggestion
            if (LoggerMethodNames.Contains(invocation.GetName())
```
Using the collection `Contains` instead of the `IEnumerable.Any` is faster.